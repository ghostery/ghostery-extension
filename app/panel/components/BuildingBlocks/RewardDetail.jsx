/**
 * Reward Detail Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import { computeTimeDelta } from '../../utils/utils';
import { sendMessage } from '../../utils/msg';
import Tooltip from '../Tooltip';

/**
 * @class Implements the details for a single reward for for the Rewards Panel
 * @memberof PanelBuildingBlocks
 */
class RewardDetail extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			copyText: t('rewards_copy_code'),
		};

		// Event Bindings
		this.handleCopyClick = this.handleCopyClick.bind(this);
		this.handleRedeemClick = this.handleRedeemClick.bind(this);
	}

	/**
	 * Lifecycle event
	 */
	componentWillMount() {
		this.props.actions.setOfferRead(this.props.id);
	}

	/**
	 * Handles clicking the copy button
	 */
	handleCopyClick() {
		// Copy the reward code
		this.copyNode.querySelector('input').select();
		document.execCommand('copy');

		// Show a toast notification
		this.props.actions.showNotification({
			text: t('rewards_code_copied_toast_notification'),
			classes: 'purple',
		});

		// Update and reset Copy Code text
		this.setState({ copyText: t('rewards_code_copied') });
		setTimeout(() => {
			this.setState({ copyText: t('rewards_copy_code') });
		}, 3000);

		// Send a signal to the offers black box
		this.props.actions.sendSignal('code_copied', this.props.id);
	}

	/**
	 * Handles clicking the redeem button
	 * @param  {Object} event the event object
	 */
	handleRedeemClick(event) {
		event.preventDefault();
		this.props.actions.sendSignal('offer_ca_action', this.props.id);
		sendMessage('openNewTab', {
			url: this.props.redeemUrl,
			become_active: true,
		});
		window.close();
	}

	/**
	 * Helper render function for the expires text.
	 * @return {JSX} JSX for the Rewards Detail
	 */
	renderExpiresText() {
		const { expires } = this.props;
		const delta = computeTimeDelta(new Date(expires), new Date());
		return t('rewards_expires_in', [delta.count, t(`rewards_expires_in_${delta.type}`)]);
	}

	/**
	 * React's required render function. Returns JSX
	 * @return {JSX} JSX for rendering the details for a Reward
	 */
	render() {
		const {
			code,
			text,
			description,
			conditions,
			logoUrl,
			pictureUrl,
			redeemText
		} = this.props;

		return (
			<div className="RewardDetail flex-container flex-dir-column align-justify">
				<div className="RewardDetail__image_container">
					<img className="RewardDetail__logo" src={logoUrl} />
					<div className="RewardDetail__picture" style={{ backgroundImage: `url('${pictureUrl}')` }} />
				</div>
				{text && (
					<div className="RewardDetail__title">
						{ text }
					</div>
				)}
				{description && (
					<div className="RewardDetail__description">
						{ description }
					</div>
				)}
				{code && (
					<div className="RewardDetail__code_container flex-container align-middle align-justify">
						<span className="RewardDetail__code" ref={(node) => { this.copyNode = node; }}>
							<span>{ code }</span>
							<input readOnly type="text" value={code} />
						</span>
						<span className="RewardDetail__copy clickable" onClick={this.handleCopyClick}>
							{this.state.copyText}
						</span>
					</div>
				)}
				<div className="RewardDetail__details_container flex-container align-justify align-middle">
					<div className="RewardDetail__expires">
						{ this.renderExpiresText() }
					</div>
					{conditions && (
						<div className="g-tooltip clickable">
							<span className="RewardDetail__terms">
								{ t('rewards_terms_conditions') }
							</span>
							<Tooltip header={conditions} position="top" delay="0" theme="dark" />
						</div>
					)}
				</div>
				<div className="RewardDetail__redeem_button clickable" onClick={this.handleRedeemClick}>
					{ redeemText }
				</div>
			</div>
		);
	}
}

export default RewardDetail;
