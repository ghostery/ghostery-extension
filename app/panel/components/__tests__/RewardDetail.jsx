/**
 * Reward Detail Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2019 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import RewardDetail from '../BuildingBlocks/RewardDetail';

// Fake the translation function to only return the translation key
global.t = function (str) {
	return str;
};

// Fake the Tooltip implementation
jest.mock('../Tooltip');

describe('app/panel/components/BuildingBlocks/RewardDetail.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('reward detail is rendered with all values present', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				code: 'test_code',
				conditions: 'test reward conditions',
				description: 'test reward description',
				expires: 1527880170315,
				id: 'test_reward_id',
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				pictureUrl: 'https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png',
				redeemUrl: 'https://www.offer-redeem-test.com',
				redeemText: 'redeem now',
				text: 'test reward title'
			};
			const component = renderer.create(<RewardDetail {...initialState} />).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward detail is rendered with missing text', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				code: 'test_code',
				conditions: 'test reward conditions',
				description: 'test reward description',
				expires: 1527880170315,
				id: 'test_reward_id',
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				pictureUrl: 'https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png',
				redeemUrl: 'https://www.offer-redeem-test.com',
				redeemText: 'redeem now',
				text: ''
			};
			const component = renderer.create(<RewardDetail {...initialState} />).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward detail is rendered with missing description', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				code: 'test_code',
				conditions: 'test reward conditions',
				description: '',
				expires: 1527880170315,
				id: 'test_reward_id',
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				pictureUrl: 'https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png',
				redeemUrl: 'https://www.offer-redeem-test.com',
				redeemText: 'redeem now',
				text: 'test reward title'
			};
			const component = renderer.create(<RewardDetail {...initialState} />).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward detail is rendered with missing code', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				code: '',
				conditions: 'test reward conditions',
				description: 'test reward description',
				expires: 1527880170315,
				id: 'test_reward_id',
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				pictureUrl: 'https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png',
				redeemUrl: 'https://www.offer-redeem-test.com',
				redeemText: 'redeem now',
				text: 'test reward title'
			};
			const component = renderer.create(<RewardDetail {...initialState} />).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward detail is rendered with missing conditions', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				code: 'test_code',
				conditions: '',
				description: 'test reward description',
				expires: 1527880170315,
				id: 'test_reward_id',
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				pictureUrl: 'https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png',
				redeemUrl: 'https://www.offer-redeem-test.com',
				redeemText: 'redeem now',
				text: 'test reward title'
			};
			const component = renderer.create(<RewardDetail {...initialState} />).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('reward detail is rendered correctly the code is copied', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				code: 'test_code',
				conditions: 'test reward conditions',
				description: 'test reward description',
				expires: 1527880170315,
				id: 'test_reward_id',
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				pictureUrl: 'https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png',
				redeemUrl: 'https://www.offer-redeem-test.com',
				redeemText: 'redeem now',
				text: 'test reward title'
			};
			const component = shallow(<RewardDetail {...initialState} />);
			expect(component.text()).toEqual(expect.stringContaining('rewards_copy_code'));
			expect(component.text()).not.toEqual(expect.stringContaining('rewards_code_copied'));
			component.setState({ copyText: t('rewards_code_copied') });
			expect(component.text()).not.toEqual(expect.stringContaining('rewards_copy_code'));
			expect(component.text()).toEqual(expect.stringContaining('rewards_code_copied'));
		});
	});
});
