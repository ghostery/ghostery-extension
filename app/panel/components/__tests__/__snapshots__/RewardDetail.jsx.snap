// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel/components/BuildingBlocks/RewardDetail.jsx Snapshot tests with react-test-renderer reward detail is rendered with all values present 1`] = `
<div
  className="RewardDetail flex-container flex-dir-column align-justify"
>
  <div
    className="RewardDetail__image_container"
  >
    <img
      className="RewardDetail__logo"
      src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
    />
    <div
      className="RewardDetail__picture"
      style={
        Object {
          "backgroundImage": "url('https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png')",
        }
      }
    />
  </div>
  <div
    className="RewardDetail__title"
  >
    test reward title
  </div>
  <div
    className="RewardDetail__description"
  >
    test reward description
  </div>
  <div
    className="RewardDetail__code_container flex-container align-middle align-justify"
  >
    <span
      className="RewardDetail__code"
    >
      <span>
        test_code
      </span>
      <input
        readOnly={true}
        type="text"
        value="test_code"
      />
    </span>
    <span
      className="RewardDetail__copy clickable"
      onClick={[Function]}
    >
      rewards_copy_code
    </span>
  </div>
  <div
    className="RewardDetail__details_container flex-container align-justify align-middle"
  >
    <div
      className="RewardDetail__expires"
    >
      rewards_expires_in
    </div>
    <div
      className="g-tooltip clickable"
    >
      <span
        className="RewardDetail__terms"
      >
        rewards_terms_conditions
      </span>
    </div>
  </div>
  <div
    className="RewardDetail__redeem_button clickable"
    onClick={[Function]}
  >
    redeem now
  </div>
</div>
`;

exports[`app/panel/components/BuildingBlocks/RewardDetail.jsx Snapshot tests with react-test-renderer reward detail is rendered with missing code 1`] = `
<div
  className="RewardDetail flex-container flex-dir-column align-justify"
>
  <div
    className="RewardDetail__image_container"
  >
    <img
      className="RewardDetail__logo"
      src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
    />
    <div
      className="RewardDetail__picture"
      style={
        Object {
          "backgroundImage": "url('https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png')",
        }
      }
    />
  </div>
  <div
    className="RewardDetail__title"
  >
    test reward title
  </div>
  <div
    className="RewardDetail__description"
  >
    test reward description
  </div>
  
  <div
    className="RewardDetail__details_container flex-container align-justify align-middle"
  >
    <div
      className="RewardDetail__expires"
    >
      rewards_expires_in
    </div>
    <div
      className="g-tooltip clickable"
    >
      <span
        className="RewardDetail__terms"
      >
        rewards_terms_conditions
      </span>
    </div>
  </div>
  <div
    className="RewardDetail__redeem_button clickable"
    onClick={[Function]}
  >
    redeem now
  </div>
</div>
`;

exports[`app/panel/components/BuildingBlocks/RewardDetail.jsx Snapshot tests with react-test-renderer reward detail is rendered with missing conditions 1`] = `
<div
  className="RewardDetail flex-container flex-dir-column align-justify"
>
  <div
    className="RewardDetail__image_container"
  >
    <img
      className="RewardDetail__logo"
      src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
    />
    <div
      className="RewardDetail__picture"
      style={
        Object {
          "backgroundImage": "url('https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png')",
        }
      }
    />
  </div>
  <div
    className="RewardDetail__title"
  >
    test reward title
  </div>
  <div
    className="RewardDetail__description"
  >
    test reward description
  </div>
  <div
    className="RewardDetail__code_container flex-container align-middle align-justify"
  >
    <span
      className="RewardDetail__code"
    >
      <span>
        test_code
      </span>
      <input
        readOnly={true}
        type="text"
        value="test_code"
      />
    </span>
    <span
      className="RewardDetail__copy clickable"
      onClick={[Function]}
    >
      rewards_copy_code
    </span>
  </div>
  <div
    className="RewardDetail__details_container flex-container align-justify align-middle"
  >
    <div
      className="RewardDetail__expires"
    >
      rewards_expires_in
    </div>
    
  </div>
  <div
    className="RewardDetail__redeem_button clickable"
    onClick={[Function]}
  >
    redeem now
  </div>
</div>
`;

exports[`app/panel/components/BuildingBlocks/RewardDetail.jsx Snapshot tests with react-test-renderer reward detail is rendered with missing description 1`] = `
<div
  className="RewardDetail flex-container flex-dir-column align-justify"
>
  <div
    className="RewardDetail__image_container"
  >
    <img
      className="RewardDetail__logo"
      src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
    />
    <div
      className="RewardDetail__picture"
      style={
        Object {
          "backgroundImage": "url('https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png')",
        }
      }
    />
  </div>
  <div
    className="RewardDetail__title"
  >
    test reward title
  </div>
  
  <div
    className="RewardDetail__code_container flex-container align-middle align-justify"
  >
    <span
      className="RewardDetail__code"
    >
      <span>
        test_code
      </span>
      <input
        readOnly={true}
        type="text"
        value="test_code"
      />
    </span>
    <span
      className="RewardDetail__copy clickable"
      onClick={[Function]}
    >
      rewards_copy_code
    </span>
  </div>
  <div
    className="RewardDetail__details_container flex-container align-justify align-middle"
  >
    <div
      className="RewardDetail__expires"
    >
      rewards_expires_in
    </div>
    <div
      className="g-tooltip clickable"
    >
      <span
        className="RewardDetail__terms"
      >
        rewards_terms_conditions
      </span>
    </div>
  </div>
  <div
    className="RewardDetail__redeem_button clickable"
    onClick={[Function]}
  >
    redeem now
  </div>
</div>
`;

exports[`app/panel/components/BuildingBlocks/RewardDetail.jsx Snapshot tests with react-test-renderer reward detail is rendered with missing text 1`] = `
<div
  className="RewardDetail flex-container flex-dir-column align-justify"
>
  <div
    className="RewardDetail__image_container"
  >
    <img
      className="RewardDetail__logo"
      src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
    />
    <div
      className="RewardDetail__picture"
      style={
        Object {
          "backgroundImage": "url('https://www.ghostery.com/wp-content/uploads/2017/12/simple-detailed-1024x833.png')",
        }
      }
    />
  </div>
  
  <div
    className="RewardDetail__description"
  >
    test reward description
  </div>
  <div
    className="RewardDetail__code_container flex-container align-middle align-justify"
  >
    <span
      className="RewardDetail__code"
    >
      <span>
        test_code
      </span>
      <input
        readOnly={true}
        type="text"
        value="test_code"
      />
    </span>
    <span
      className="RewardDetail__copy clickable"
      onClick={[Function]}
    >
      rewards_copy_code
    </span>
  </div>
  <div
    className="RewardDetail__details_container flex-container align-justify align-middle"
  >
    <div
      className="RewardDetail__expires"
    >
      rewards_expires_in
    </div>
    <div
      className="g-tooltip clickable"
    >
      <span
        className="RewardDetail__terms"
      >
        rewards_terms_conditions
      </span>
    </div>
  </div>
  <div
    className="RewardDetail__redeem_button clickable"
    onClick={[Function]}
  >
    redeem now
  </div>
</div>
`;
